### ISSUE TITLE
Reduce min-value for task_tracker_delay.

### ISSUE THREAD

**Issue by sumedhpathak** [citusdata/citus#826]
Currently task_tracker_delay can be reduced to at most 10ms. Change it to allow lower values (1ms).

**Comment by metdos** [comment]
I run some tests to see effects of task_tracker_delay value. I run my tests on EC2 on c3.4xlarge instances. I used tpch with scale 1. I used first three queries from multi_single_relation_subquery.sql in regression tests.

|64 Shards (1 + 3 nodes)	|1ms|	10ms	|200ms|
| :--- | :---  |  :---  | :---  |
|Query1	|6.8	|7.1	|15.9|
|Query2	|6.9	|7.1	|15.7|
|Query3	|8.2 |8.4	|19.1|

|256 Shards (1 + 3 nodes)	|1ms|	10ms|	200ms|
| :--- | :---  |  :---  | :---  |
|Query1|	29.0|	29.3|	60.8|
|Query2|	29.1	|29.4|	60.6|
|Query3|	30.4	|30.9|	63.7|
			
			
|256 Shards (1 + 5 nodes)|	1ms	|10ms|	200ms|
| :--- | :---  |  :---  | :---  |
|Query1|	30.8	|30.5|61.1|
|Query2|	30.4	|30.4|59.8|
|Query3|	35.2|35.1	|66.4|

**Comment by metdos** [comment]
@sumedhpathak, @ozgune, I have two questions;

1. Looking into test results, I couldn't see any huge effects of using 1ms instead of 10ms. Still, we can reduce min value to 1ms to eliminate possible questions. Any comments?

2. Do you have any setup suggestions where we can see effects of using 1ms?

**Comment by ozgune** [comment]
@metdos, I'm inlning my notes below.

>> Looking into test results, I couldn't see any huge effects of using 1ms instead of 10ms. Still, we can reduce min value to 1ms to eliminate possible questions. Any comments?

I'm inclined to reduce the minimum value to 1ms. This change won't have an adverse effect on production deployments; and allows us to selectively try reducing this value if we'd like to.

>> Do you have any setup suggestions where we can see effects of using 1ms?

I have two suggestions. What if we tried the following to see if it helps?

1. In production settings, we usually reduce ``remote_task_check_interval`` to 1-5ms.
2. This change may have more of an impact when we have concurrent queries. In that case, query fragments will get queued on workers. What if we tried running 64 shards with 5 or 10 concurrent queries? (and increased the max_tracked_tasks setting to allow for more queuing)

**Comment by metdos** [comment]
I run some more tests to see the effects of concurrent queries and max_running_tasks_per_node.

Here are results (X: query count, ms value is for task_tracker_delay, "-" range results are in seconds):

| max_running_tasks_per_node| 1X - 1ms| 5X - 1ms| 10X- 1ms| 10X - 10ms| 10X - 10ms| 
| :--- | :---  |  :---  | :---  |:---  |:---  |
| 8	| 2.2	| 6-7| 15-16| 10-10| 	22-23| 
| 32	| 7.6	| 17-17	| 43-43| 	 16-18| 50-50| 
| 64	| 19.5	| 38-39	| 61-70	| 38-40	| 90-100|

**Summary:**

1. Lower task_tracker_delay value (1ms) is especially effective if there are concurrent queries (10X). I'm opening a PR to reduce the minimum limit.

2. Increasing max_running_tasks_per_node actually hurts the performance, at least in my tests. I checked different things, but I could not pinpoint the problem. This needs more research. For single query, I checked these items;

- There is a time frame in the middle of the query where there is no IO, almost no cpu utilization, no active query other than task_tracker_task_status and no locks in pg_locks.
- Disk read is not bottleneck.
- I increased shared_buffer to 5GB, it didn't help.

@sumedhpathak, @ozgune any ideas? Should I open a separate issue to investigate this?

Some notes about test setup:
(2 + 1) EC2 c3.4xlarge 
64 shards
tpch with scale 1
the first query from multi_single_relation_subquery.sql 
remote_task_check_interval to 1ms

How to set guc values:

```bash
ALTER SYSTEM SET citus.max_running_tasks_per_node = 64;
ALTER SYSTEM SET citus.task_tracker_delay = '1ms';
SELECT pg_reload_conf();
```

Check guc values:
```bash
SHOW citus.max_running_tasks_per_node;
SHOW citus.task_tracker_delay;
```

Run queries in parallel:

```bash
for i in {1..10}
do
  time /usr/local/pgsql/bin/psql -e -d postgres -c "SET citus.task_executor_type TO 'task-tracker';
  select
      number_sum,
      count(*) as total,
      avg(total_count) avg_count
  from
      (select
          l_suppkey,
          l_linestatus,
          sum(l_linenumber) as number_sum,
          count(*) as total_count
      from
          lineitem
      group by
          l_suppkey,
          l_linestatus) as distributed_table
  where
      number_sum >= 10
  group by
      number_sum
  order by
      total desc,
      number_sum desc
  limit 10;" > test_output.log &
  sleep 0.1s
done
```

**Comment by sumedhpathak** [comment]
@metdos Do you know how many tasks each re-partition query generates? What was the value for max_tracked_tasks_per_node? Does increasing that number impact the performance in anyway? 
I'm curious how many tasks get assigned to a worker node when you have say 10 concurrent similar queries, and how many are runnable (or start running). Could you spend an hour or two tomorrow morning looking into that?

Finally, should we lower the default as well, as part of this change?

**Comment by metdos** [comment]
@sumedhpathak, @ozgune,

I found the reason of why increasing max_running_tasks_per_node hurts the performance. We use max_running_tasks_per_node in [hash partition count calculation](https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/multi_physical_planner.c#L1765), and increasing max_running_tasks_per_node increases the number of tasks created.

I set up a similar 2 + 1 node cluster above and verified that running 10 queries in parallel with max_running_tasks_per_node=64 takes around 61-70 seconds. To test my hypothesis, I created a test branch and [returned a fixed number "8" for hash partition count](https://github.com/citusdata/citus/commit/a0e1c280a81dd1f36008784f524381dc3d4264af). Running this version of citus, decreased the run time for 10 parallel queries **from 61-70 seconds to 8-9 seconds** (max_running_tasks_per_node is still 64, but hash partition count is independent of it).

Inferring the optimum hash partition count for every use-case does not seem realistic. Therefore, 
my proposal is to create another guc variable which sets hash partition count for power users. What do you think?

For other questions:

**i. How many tasks:**
For 10 queries with max_running_tasks_per_node=64, 21440 tasks per worker node.

Here is how to calculate total tasks per query:

```
MAP_TASKs: shardCount + 
MAP_OUTPUT_FETCH_TASKs: shardCount * partitionCount + 
MERGE_TASKs: partitionCount + 
SQL_TASKs: partitionCount
```

Shell script calculate total tasks per query:

```bash
shardCount=64
nodeCount=2
MaxRunningTasksPerNode=64
let partitionCount="$nodeCount * ($MaxRunningTasksPerNode/2)"

let totalTaskCount="$shardCount + ($shardCount * $partitionCount) + $partitionCount + $partitionCount"
echo $totalTaskCount
```

**ii. Effects of max_tracked_tasks_per_node:**

This only sets the maximum number of tasks which can be assigned. If we set this low, then we get out of shared memory errors. This setting is used in required shared memory size estimation, so unnecessarily setting this high can cause unnecessarily to create a huge task tracker hash table in shared memory. I didn't see any effects of this variable on the query performance.

**iii. Shall we lower the default of task_tracker_delay as well:**

I also considered this. However, running regression tests locally, time to time I get warnings related to 10ms task_tracker_delay. So, I though it should be okay to leave it high for the default case. Maybe we can decrease it to a more conservative value such as 50-100ms. What do you think?

**Comment by sumedhpathak** [comment]
@metdos I tested the changes, however I found the biggest performance difference actually comes from modifying the hash-partition-count via the master.

With 10ms as the task_tracker_delay:
With a value of 16:
Single task: ~4.1 seconds
Parallel tasks: ~20-21 seconds each.

With a value of 8:
Single task: ~2.4 seconds
Parallel tasks: ~14.5 seconds each.

I didn't see much variation between 1ms and 10ms task-tracker-delay, so that's interesting as well.
Worker node settings:
```
citus.task_tracker_delay=10ms
citus.max_running_tasks_per_node=64
citus.max_tracked_tasks_per_node=50000
citus.remote_task_check_interval=1ms

```

Master node settings:
```
citus.max_running_tasks_per_node=8
citus.remote_task_check_interval=1ms
citus.max_tracked_tasks_per_node=50000
citus.task_executor_type='task-tracker'

```

**Comment by sumedhpathak** [comment]
Also, I tested on 2+1 m3.2xlarge instances. Do you see similar behavior? If so, we could just recommend reducing max_running_tasks_per_node on the master for an immediate fix.

**Comment by metdos** [comment]
Hey @sumedhpathak 

> I didn't see much variation between 1ms and 10ms task-tracker-delay, so that's interesting as well.

Ohh :S, this is interesting. Looking to my results above for max_running_tasks_per_node=8 on 10 concurrent queries, there is an effect of 1ms task_tracker_delay. Therefore, I rerun some of my tests (results are below). **Interestingly, increasing the value of max_running_tasks_per_node on worker nodes decreases the effect of having task_tracker_delay as 1ms.** We can recommend reducing max_running_tasks_per_node on the master for an immediate fix. Actually on my setup reducing max_running_tasks_per_node to 4 or 2 also had a positive effect. However, having task_tracker_delay as 1ms still can have some effects on different workloads, and I think we should make that change.

Here are the test results;

The master node settings are fixed:

```
citus.max_running_tasks_per_node=8;
citus.remote_task_check_interval='1ms';
citus.max_tracked_tasks_per_node=50000;
citus.task_executor_type='task-tracker';
```

Fixed worker node settings:
```
citus.remote_task_check_interval='1ms';
citus.max_tracked_tasks_per_node=50000;
```

Results of changing values of max_running_tasks_per_node and task_tracker_delay on only worker nodes can be seen on the table.

|max_running_tasks_per_node of worker nodes|	1X-1ms|	1X-10ms|	10X-1ms|	10X-10ms|
| :--- | :---  |  :---  | :---  |:---  |:---  |
|8	|1.9|	2.3|	13-14|	20-21|
|16|	1.9|	2.1	|9-10	|12-13|
|64|	2|	2.2|	9-10|	9-10|


For 10 concurrent queries with max_running_tasks_per_node=64 and task_tracker_delay=10ms on worker nodes, queries finish in about 9-10 seconds. Without changing the worker node settings, if we reduce max_running_tasks_per_node on master node to 4 and 2, query run times drop to 7-8 seconds and 6-7 seconds respectively.
